### DISCUSSION TITLE
basic exemple somewhere ?

### DISCUSSION THREAD

**Discussion by benoitmercusot** [dnadoba/MIDIKit#9] [Q&A] [answered]
Hi, i'm very intersting with your library but since i'm new to Swift i'm unable to find a working exemple for now.

From the doc :

do {
    let outputPort = try midiClient.makeOutputPort(name: "Output") 
    outputPort.send(MIDIMessage.controlChange(channel: 1, controller: 11, value: 64), to: destination)
} catch {
    // catch error
}

what should be the "destination" variable.
for now, how i'm supposed to send note to an output of my choice ?

Thank you very much for your help.
Benoit.

**Comment by dnadoba** [comment, marked as answer]
Hi Benoit,

You can find a working example in the Playground that is in the root folder of this repository. I have just update that playground.
When you open MIDIKit in Xcode, you can just click on the Playground and you should be able to run it.
It will print all currently connected devices and will start listening for the incoming MIDI messages. Make sure your MIDI Device is connected to your Mac before you execute the Playground.

Here is the code of the Playground:
```swift
import Foundation
import MIDIKit
import PlaygroundSupport

// Keep playground running
PlaygroundPage.current.needsIndefiniteExecution = true

let client = MIDIClient(name: "MidiTestClient")
try client.start()

// log all currently connected devices
print("Devices:")
MIDIDevice.getAll().forEach {
    print("   \($0.name as Any)")
    print("   Entities:")
    $0.getEntities().forEach {
        print("      \($0.name as Any)")
        print("      Sources:")
        $0.getSources().forEach {
            print("         \($0.name as Any)")
        }
        print("      Destinations:")
        $0.getDestinations().forEach {
            print("         \($0.name as Any)")
        }
    }
}
// receiving all messages
let inputPorts = try MIDIEndpoint.getAllSources().map { source -> MIDIInputPort in
    let inputPort = try client.makeInputPort(name: "MIDItest_inputPort") { (result) in
        do {
            let packet = try result.get()
            print("MIDI Received From Source: \(packet.source.displayName ?? "-") \(packet.source.entity?.device?.identifier.debugDescription ?? "-")")
            for message in packet.messages {
                print(message)
            }
        } catch {
            print(error)
        }
    }
    try inputPort.connect(to: source)
    return inputPort
}



let outputPort = try client.makeOutputPort(name: "MIDITest_outputPort")

/// sends message to all devices
/// - Parameter messages: MIDI messages to send
func send(_ messages: [MIDIMessage]) throws {
    let allDesitnations = MIDIEndpoint.getAllDestinations()
    if allDesitnations.isEmpty {
        print("no destinations available")
    }
    try allDesitnations.forEach { (destination) in
        try outputPort.send(messages, to: destination)
    }
}

/// send messages to all devices
try send([.controlChange(channel: 0, controller: 0, value: 0)])
```

To come back to your initial question:
`destination` must be a [`MIDIEndpoint`](https://dnadoba.github.io/MIDIKit/Structs/MIDIEndpoint.html) which you can get by calling [`MIDIEndpoint.getAllDestinations()`](https://dnadoba.github.io/MIDIKit/Structs/MIDIEndpoint.html#/s:7MIDIKit12MIDIEndpointV18getAllDestinationsSayACGyFZ) or if you have a `MIDIEntity` you can call [`.getDestinations()`](https://dnadoba.github.io/MIDIKit/Structs/MIDIEntity.html#/s:7MIDIKit10MIDIEntityV15getDestinationsSayAA12MIDIEndpointVGyF) on it.
You can also be notified of new Devices/Endpoints by implementing the [`MIDIClientDelegate`](https://dnadoba.github.io/MIDIKit/Protocols/MIDIClientDelegate.html#/s:7MIDIKit18MIDIClientDelegateP10midiClient_10didRecieveyAA0B0C_AG12NotificationOtF).

You can find more Information about all available types and their methods/properites by looking at the documentation at [https://dnadoba.github.io/MIDIKit/index.html](https://dnadoba.github.io/MIDIKit/index.html).
Their is currently no real documentation but you can get an overview of the available types and how they relate.

Hope that helps :)

**Reply by benoitmercusot** [reply]
Very usefull ! 
Thank you :)
